src/shared/group-record.c

```diff
@@ -0,0 +1,347 @@
+                { "status",         JSON_VARIANT_OBJECT,   NULL,                           0,                                       0         },
+
+                /* Ignore 'signature', we check it with explicit accessors instead */
+                { "signature",      JSON_VARIANT_ARRAY,    NULL,                           0,                                       0          },
+                {},
+        };
+
+        JsonDispatchFlags json_flags = USER_RECORD_LOAD_FLAGS_TO_JSON_DISPATCH_FLAGS(load_flags);
+        JsonVariant *per_machine, *binding, *status;
+        int r;
+
+        assert(h);
+        assert(!h->json);
+
+        /* Note that this call will leave a half-initialized record around on failure! */
+
+        if ((USER_RECORD_REQUIRE_MASK(load_flags) & (USER_RECORD_SECRET|USER_RECORD_PRIVILEGED)) != 0)
+                return json_log(v, json_flags, SYNTHETIC_ERRNO(EINVAL), "Secret and privileged section currently not available for groups, refusing.");
+
+        r = user_group_record_mangle(v, load_flags, &h->json, &h->mask);
+        if (r < 0)
+                return r;
+
+        r = json_dispatch(h->json, group_dispatch_table, NULL, json_flags, h);
+        if (r < 0)
+                return r;
+
+        /* During the parsing operation above we ignored the 'perMachine' and 'binding' fields, since we want
+         * them to override the global options. Let's process them now. */
+
+        per_machine = json_variant_by_key(h->json, "perMachine");
+        if (per_machine) {
+                r = dispatch_per_machine("perMachine", per_machine, json_flags, h);
+                if (r < 0)
+                        return r;
+        }
+
+        binding = json_variant_by_key(h->json, "binding");
+        if (binding) {
+                r = dispatch_binding("binding", binding, json_flags, h);
+                if (r < 0)
+                        return r;
+        }
+
+        status = json_variant_by_key(h->json, "status");
+        if (status) {
+                r = dispatch_status("binding", status, json_flags, h);
+                if (r < 0)
+                        return r;
+        }
+
+        if (FLAGS_SET(h->mask, USER_RECORD_REGULAR) && !h->group_name)
+                return json_log(h->json, json_flags, SYNTHETIC_ERRNO(EINVAL), "Group name field missing, refusing.");
+
+        r = group_record_augment(h, json_flags);
+        if (r < 0)
+                return r;
+
+        return 0;
+}
+
+int group_record_build(GroupRecord **ret, ...) {
+        _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
+        _cleanup_(group_record_unrefp) GroupRecord *g = NULL;
+        va_list ap;
+        int r;
+
+        assert(ret);
+
+        va_start(ap, ret);
+        r = json_buildv(&v, ap);
+        va_end(ap);
+
+        if (r < 0)
+                return r;
+
+        g = group_record_new();
+        if (!g)
+                return -ENOMEM;
+
+        r = group_record_load(g, v, USER_RECORD_LOAD_FULL);
```

Comment:
we actually use the json object all the time. The idea is that GroupRecord always contains the parsed version and the JSON version so that we can use either depending on the context. The calls that update the user record are careful to update both too.

Basically, whenever we process user records internally we want the parsed version but we quite often want to send the things over varlink or write it to disk, and then we want the json variant.